Use `git config credential.helper 'store --file=/tmp/file' for `git push`

Ocramius opened the pull request:
This should fix some underlying design issues described in #198
    
The full URI of the repository, which includes a token in plaintext, is now no
longer the actual `git remote` URI, and is instead stored in a temporary
file which `git` uses to authenticate with the remote.
    
This prevents the remote URI from leaking in `STDOUT`/`STDERR`, which are
visible in case of crashes.

This is a **partial** fix of #198 - the issue with `git` being broken upfront is still there.

Ocramius committed:
Use `git config credential.helper 'store --file=/tmp/file' for `git push`

This should fix some underlying design issues described in #198

The full URI of the repository, which includes a token in plaintext, is now no
longer the actual `git remote` URI, and is instead stored in a temporary
file which `git` uses to authenticate with the remote.

This prevents the remote URI from leaking in `STDOUT`/`STDERR`, which are
visible in case of crashes.

weierophinney commented on src/Application/Command/BumpChangelogForReleaseBranch.php:
```diff
@@ -44,12 +44,15 @@ public function execute(InputInterface $input, OutputInterface $output): int
         Psl\invariant(Filesystem\is_directory($repositoryPath . '/.git'), 'Workspace is not a GIT repository.');
 
-        ($this->fetch)($repositoryCloneUri, $repositoryPath);
+        ($this->fetch)(
+            $repositoryName->uri(),
+            $repositoryName->uriWithTokenAuthentication($this->environment->githubToken()),
+            $repositoryPath
+        );
```
Trying to understand how this fixes the problem - is it because in the case of a crash, only the first argument (here, `$repositoryName->uri()`, previously, a version of the URI with the token included) is ever logged?

Ocramius replied:
It was in the exception message.

I think a validating test case is needed.

Ocramius committed:
Test: failing to fetch a repository and create credentials does NOT exfiltate those credentials

This test fails on `1.14.x`, and succeeds on `HEAD`, therefore validating the current patch, and
preventing a security regression.

Ocramius commented on test/unit/Git/FetchAndSetCurrentUserByReplacingCurrentOriginRemoteTest.php:
```diff
@@ -105,4 +106,27 @@ public function testFetchesAndSetsCurrentUserWithoutOrigin(): void
         self::assertStringContainsString('origin/new-branch', $fetchedBranches);
     }
+
+    public function testFailingToSetARemoteWillNotLeadToSecretExfiltration(): void
+    {
+        $sourceUri            = $this->createMock(UriInterface::class);
+        $sourceUriWithSecrets = $this->createMock(UriInterface::class);
+
+        $sourceUri->method('__toString')
+            ->willReturn('https://github.com/laminas/this-repository-does-not-exist.git');
+        $sourceUriWithSecrets->method('__toString')
+            ->willReturn('https://SUPERSECRET:x-oauth-basic@github.com/laminas/this-repository-does-not-exist.git');
+
+        $notAGitDirectory = Filesystem\create_temporary_file();
+
+        Filesystem\delete_file($notAGitDirectory);
+        Filesystem\create_directory($notAGitDirectory);
+
+        try {
+            (new FetchAndSetCurrentUserByReplacingCurrentOriginRemote($this->variables))
+                ->__invoke($sourceUri, $sourceUriWithSecrets, $notAGitDirectory);
+        } catch (Throwable $failure) {
+            self::assertDoesNotMatchRegularExpression('/SUPERSECRET/m', $failure->getMessage());
+        }
+    }
```
Verified locally that this test fails on `1.14.x`, while succeeds on this branch.
